File: src/main/kotlin/Character.kt
import utils.ByteBufferBackedInputStream
import utils.LittleEndianDataInputStream
import java.io.ByteArrayOutputStream
import java.io.File
import java.io.DataInputStream
import java.io.ObjectOutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.channels.FileChannel
import java.nio.file.Files


private val logger = mu.KotlinLogging.logger {}

class Character constructor(val caracterName:String, val caracterFile:String) {
private val  beginBlockPattern = byteArrayOf(

  11,
  0,
  0,
  0,
  98,
  101,
  103,
  105,
  110,
  95,
  98,
  108,
  111,
  99,
  107
)
private val endBlockPattern = byteArrayOf(

  9,
  0,
  0,
  0,
  101,
  110,
  100,
  95,
  98,
  108,
  111,
  99,
  107
)
//private var rawData:ByteArray
private var itemBlockStart:Int =0
private var itemBlockEnd:Int =0
private var equipmentBlockStart:Int =0
private var equipmentBlockEnd:Int =0
private var numberOfSacks:Int =0
private var currentlyFocusedSackNumber:Int =0
private var currentlySelectedSackNumber:Int=0
private var equipmentCtrlIOStreamVersion:Int=0
private val sacks:MutableList<Sack> = mutableListOf()
private var equipmentSack:Sack = Sack()
public var isModified = false
  private var rawData:ByteBuffer = ByteBuffer.allocate(0)



public fun loadFile():Unit
    {
      val file = File(caracterFile)
        try {

            rawData = FileChannel.open(file.toPath()).map(FileChannel.MapMode.READ_ONLY,0,file.length())
            rawData.order( ByteOrder.LITTLE_ENDIAN);

            this.parseRawData();
        }catch(ex:Exception){
            logger.error { "Error opening file $caracterFile ${ex.message} $ex" }
        }
    }

    private fun parseRawData():Unit
    {
          var num = 0;
          var start = 0;
          var offset1 = 0;
          var offset2 = 0;
          var flag1 = false;
          var flag2 = false;
          var nextBlockDelim = 0
          while (run { nextBlockDelim = FindNextBlockDelim( start)
              (!flag1 || !flag2) && nextBlockDelim != -1})
          {
            if ( rawData[nextBlockDelim] ==  beginBlockPattern[0])
            {
              ++num;
              val offset3 = nextBlockDelim + beginBlockPattern.size + 4;
              val upperInvariant = TQData.readCString(rawData,offset3).uppercase()
              start = rawData.position() //offset3 + upperInvariant.length +1
              var str:String
              if (upperInvariant == "BEGIN_BLOCK")
              {
                str = "(NONAME)"
                start -= beginBlockPattern.size
              }
              else if (upperInvariant == "END_BLOCK")
              {
                str = "(NONAME)"
                start -= endBlockPattern.size;
              }
              else if (upperInvariant == "ITEMPOSITIONSSAVEDASGRIDCOORDS")
              {
                start += 4;
                offset1 = start;
                flag1 = true;
              }
              else if (upperInvariant =="USEALTERNATE")
              {
                start += 4;
                offset2 = start;
                flag2 = true;
              }
            }
            else
            {
              --num;
              start = nextBlockDelim + endBlockPattern.size
            }
          }
          if (flag1)
          {
            try
            {
              this.parseItemBlock(offset1);
            }
            catch ( ex:Exception)
            {
              logger.debug("Error parsing player file Item Block - $caracterName", ex)
//              throw new ArgumentException("Error parsing player file Item Block- " + this.caracterFile, (Exception) ex);
            }
          }
          if (true)
            return;
          try
          {
            this.parseEquipmentBlock(offset2);
          }
          catch ( ex:Exception)
          {
            logger.debug("Error parsing player file Item Block - " + this.caracterName);
            logger.debug(ex.toString());
          }
        }


    private fun FindNextBlockDelim(start:Int):Int
    {
      var index1 = 0
      var index2 = 0
      for ( index3 in start until rawData.limit() )
      {
        if (rawData[index3] == beginBlockPattern[index1])
        {
          ++index1;
          if (index1 == beginBlockPattern.size)
            return index3 + 1 - index1;
        }
        else if (index1 > 0)
        {
          index1 = 0;
          if (rawData[index3] == beginBlockPattern[index1])
            ++index1;
        }
        if (rawData[index3] == endBlockPattern[index2])
        {
          ++index2;
          if (index2 == endBlockPattern.size)
            return index3 + 1 - index2;
        }
        else if (index2 > 0)
        {
          index2 = 0;
          if (rawData[index3] == endBlockPattern[index2])
            ++index2;
        }
      }
      return -1;
    }

    private fun parseEquipmentBlock(offset:Int):Unit
    {
      try
      {
        this.equipmentBlockStart = offset;
        TQData.validateNextString(rawData,"equipmentCtrlIOStreamVersion", offset)
        this.equipmentCtrlIOStreamVersion = rawData.getInt();
        this.equipmentSack = Sack()
        this.equipmentSack.sackType = SackType.Equipment;
        this.equipmentSack.IsImmortalThrone = true;
        this.equipmentSack.Parse(rawData);
        this.equipmentBlockEnd =  rawData.position()
      }
      catch (ex:Exception )
      {
        logger.error { "Error $ex" }
      }
    }

    private fun parseItemBlock(offset:Int):Unit
    {
      try
      {
        this.itemBlockStart = offset;
        TQData.validateNextString(rawData, "numberOfSacks", offset);
        this.numberOfSacks = rawData.getInt();
        TQData.validateNextString(rawData,"currentlyFocusedSackNumber");
        this.currentlyFocusedSackNumber = rawData.getInt();
        TQData.validateNextString(rawData,"currentlySelectedSackNumber");
        this.currentlySelectedSackNumber = rawData.getInt();
//        sacks. = Sack[this.numberOfSacks];
        for (index in 0 until numberOfSacks)
        {
          val sack =  Sack();
          sack.sackType = SackType.Sack;
          sack.IsImmortalThrone = true;
          sack.Parse(rawData);
          sacks[index] = sack
        }
        this.itemBlockEnd = rawData.position()
      }
      catch ( ex:java.lang.Exception)
      {
        throw ex
      }
    }

    public fun CreateEmptySacks():Sack
    {
      isModified = false
      return Sack()

    }

    public fun emptyMainSack()
    {
      GetMainSack().EmptySack();
      isModified = true;
    }

    public fun Save( fileName:String, backup: Boolean = false):Unit
    {
      var str1 = "";
      if (backup)
      {
        val orig = File(fileName)
        val dir = orig.parentFile
        val str2 = File(dir, "BlaksmithBackup_Player.chr")
        if (!str2.exists())
        {
          logger.info("Save file backup created: $str2")
          Files.copy(orig.toPath(), str2.toPath());
        }
        str1 = " Original file backed to BlaksmithBackup_Player.chr";
      }
      val buffer = Encode();
      Files.write(File(fileName).toPath(),buffer)
      this.isModified = false;
      this.GetMainSack().isModified = false;
      logger.info("Character data saved to $fileName $str1");
    }

     fun Encode():ByteArray
    {
      val sourceArray = this.EncodeItemData();
        val tempArray = ByteArray(rawData.limit() - this.itemBlockEnd)
      var destinationArray = ByteArray(itemBlockStart + sourceArray.size + (rawData.limit() - this.itemBlockEnd))
        rawData.get(destinationArray,0,itemBlockStart)
      sourceArray.copyInto(destinationArray, 0,  this.itemBlockStart, sourceArray.size)
      this.rawData.get(this.itemBlockEnd, destinationArray, itemBlockStart + sourceArray.size, this.rawData.limit() - this.itemBlockEnd)
      return destinationArray
    }

    private fun EncodeEquipmentData():ByteArray
    {
      var length = 0
      val output =  ByteArrayOutputStream(2048)

        val writer = ObjectOutputStream(output)
          TQData.writeCString(writer, "equipmentCtrlIOStreamVersion");
          writer.write(this.equipmentCtrlIOStreamVersion);
          this.equipmentSack.Encode(writer);
          length = output.size()

        val buffer = output.toByteArray()

      return if (length == buffer.size) {
        buffer
      } else {
        buffer.copyOfRange(0, length)
      }
    }

    private fun EncodeItemData():ByteArray
    {
      var length = 0
      val output =  ByteArrayOutputStream(2048)

      val writer = ObjectOutputStream(output)
          TQData.writeCString(writer, "numberOfSacks");
          writer.write(this.numberOfSacks);
          TQData.writeCString(writer, "currentlyFocusedSackNumber");
          writer.write(this.currentlyFocusedSackNumber);
          TQData.writeCString(writer, "currentlySelectedSackNumber");
          writer.write(this.currentlySelectedSackNumber);
          sacks.forEach { sack -> sack.Encode(writer) }

          length =  output.size()

        val buffer = output.toByteArray()

        return if (length == buffer.size) {
            buffer
        } else {
            buffer.copyOfRange(0, length)
        }
    }


    public fun GetMainSack() = sacks[0]

    public override fun equals(other: Any?): Boolean {
        return if (other is Character) {
            caracterName == other.caracterName
        } else {
            false
        }
    }

    public override fun hashCode(): Int = caracterName.hashCode()


  }

File: src/main/kotlin/done/Character.cs
// Decompiled with JetBrains decompiler
// Type: TQ_weaponsmith.Character
// Assembly: TQ_Blacksmith, Version=0.1.4.1, Culture=neutral, PublicKeyToken=null
// MVID: 1CC9040D-95C6-4C54-B248-9D80684B2CA8
// Assembly location: D:\Games\GOG Galaxy\Games\Titan Quest - Anniversary Edition\Downloads\Blacksmith\TQ_Blacksmith.exe

using System;
using System.IO;

namespace TQ_weaponsmith
{
  public class Character
  {
    private static byte[] beginBlockPattern = new byte[15]
    {
      (byte) 11,
      (byte) 0,
      (byte) 0,
      (byte) 0,
      (byte) 98,
      (byte) 101,
      (byte) 103,
      (byte) 105,
      (byte) 110,
      (byte) 95,
      (byte) 98,
      (byte) 108,
      (byte) 111,
      (byte) 99,
      (byte) 107
    };
    private static byte[] endBlockPattern = new byte[13]
    {
      (byte) 9,
      (byte) 0,
      (byte) 0,
      (byte) 0,
      (byte) 101,
      (byte) 110,
      (byte) 100,
      (byte) 95,
      (byte) 98,
      (byte) 108,
      (byte) 111,
      (byte) 99,
      (byte) 107
    };
    private string caracterName;
    private string caracterFile;
    private byte[] rawData;
    private int itemBlockStart;
    private int itemBlockEnd;
    private int equipmentBlockStart;
    private int equipmentBlockEnd;
    private int numberOfSacks;
    private int currentlyFocusedSackNumber;
    private int currentlySelectedSackNumber;
    private int equipmentCtrlIOStreamVersion;
    private Sack[] sacks;
    private Sack equipmentSack;
    public bool isModified;

    public Character(string name, string filePath)
    {
      this.caracterName = name;
      this.caracterFile = filePath;
    }

    public void loadFile()
    {
      using (FileStream input = new FileStream(this.caracterFile, FileMode.Open, FileAccess.Read))
      {
        using (BinaryReader binaryReader = new BinaryReader((Stream) input))
          this.rawData = binaryReader.ReadBytes((int) input.Length);
      }
      try
      {
        this.parseRawData();
      }
      catch (ArgumentException ex)
      {
        throw;
      }
    }

    private void parseRawData()
    {
      using (MemoryStream input = new MemoryStream(this.rawData, false))
      {
        using (BinaryReader reader = new BinaryReader((Stream) input))
        {
          int num = 0;
          int start = 0;
          int offset1 = 0;
          int offset2 = 0;
          bool flag1 = false;
          bool flag2 = false;
          int nextBlockDelim;
          while ((!flag1 || !flag2) && (nextBlockDelim = this.FindNextBlockDelim(start)) != -1)
          {
            if ((int) this.rawData[nextBlockDelim] == (int) Character.beginBlockPattern[0])
            {
              ++num;
              int offset3 = nextBlockDelim + Character.beginBlockPattern.Length + 4;
              input.Seek((long) offset3, SeekOrigin.Begin);
              string upperInvariant = TQData.readCString(reader).ToUpperInvariant();
              start = (int) input.Position;
              string str;
              if (upperInvariant.Equals("BEGIN_BLOCK"))
              {
                str = "(NONAME)";
                start -= Character.beginBlockPattern.Length;
              }
              else if (upperInvariant.Equals("END_BLOCK"))
              {
                str = "(NONAME)";
                start -= Character.endBlockPattern.Length;
              }
              else if (upperInvariant.Equals("ITEMPOSITIONSSAVEDASGRIDCOORDS"))
              {
                start += 4;
                offset1 = start;
                flag1 = true;
              }
              else if (upperInvariant.Equals("USEALTERNATE"))
              {
                start += 4;
                offset2 = start;
                flag2 = true;
              }
            }
            else
            {
              --num;
              start = nextBlockDelim + Character.endBlockPattern.Length;
            }
          }
          if (flag1)
          {
            try
            {
              this.parseItemBlock(offset1, reader);
            }
            catch (ArgumentException ex)
            {
              Logger.Debug("Error parsing player file Item Block - " + this.caracterName);
              Logger.Debug(ex.ToString());
              throw new ArgumentException("Error parsing player file Item Block- " + this.caracterFile, (Exception) ex);
            }
          }
          if (true)
            return;
          try
          {
            this.parseEquipmentBlock(offset2, reader);
          }
          catch (ArgumentException ex)
          {
            Logger.Debug("Error parsing player file Item Block - " + this.caracterName);
            Logger.Debug(ex.ToString());
            throw new ArgumentException("Error parsing player file Item Block- " + this.caracterFile, (Exception) ex);
          }
        }
      }
    }

    private int FindNextBlockDelim(int start)
    {
      int index1 = 0;
      int index2 = 0;
      for (int index3 = start; index3 < this.rawData.Length; ++index3)
      {
        if ((int) this.rawData[index3] == (int) Character.beginBlockPattern[index1])
        {
          ++index1;
          if (index1 == Character.beginBlockPattern.Length)
            return index3 + 1 - index1;
        }
        else if (index1 > 0)
        {
          index1 = 0;
          if ((int) this.rawData[index3] == (int) Character.beginBlockPattern[index1])
            ++index1;
        }
        if ((int) this.rawData[index3] == (int) Character.endBlockPattern[index2])
        {
          ++index2;
          if (index2 == Character.endBlockPattern.Length)
            return index3 + 1 - index2;
        }
        else if (index2 > 0)
        {
          index2 = 0;
          if ((int) this.rawData[index3] == (int) Character.endBlockPattern[index2])
            ++index2;
        }
      }
      return -1;
    }

    private void parseEquipmentBlock(int offset, BinaryReader reader)
    {
      try
      {
        this.equipmentBlockStart = offset;
        reader.BaseStream.Seek((long) offset, SeekOrigin.Begin);
        TQData.validateNextString("equipmentCtrlIOStreamVersion", reader);
        this.equipmentCtrlIOStreamVersion = reader.ReadInt32();
        this.equipmentSack = new Sack();
        this.equipmentSack.SackType = SackType.Equipment;
        this.equipmentSack.IsImmortalThrone = true;
        this.equipmentSack.Parse(reader);
        this.equipmentBlockEnd = (int) reader.BaseStream.Position;
      }
      catch (ArgumentException ex)
      {
        throw;
      }
    }

    private void parseItemBlock(int offset, BinaryReader reader)
    {
      try
      {
        this.itemBlockStart = offset;
        reader.BaseStream.Seek((long) offset, SeekOrigin.Begin);
        TQData.validateNextString("numberOfSacks", reader);
        this.numberOfSacks = reader.ReadInt32();
        TQData.validateNextString("currentlyFocusedSackNumber", reader);
        this.currentlyFocusedSackNumber = reader.ReadInt32();
        TQData.validateNextString("currentlySelectedSackNumber", reader);
        this.currentlySelectedSackNumber = reader.ReadInt32();
        this.sacks = new Sack[this.numberOfSacks];
        for (int index = 0; index < this.numberOfSacks; ++index)
        {
          this.sacks[index] = new Sack();
          this.sacks[index].SackType = SackType.Sack;
          this.sacks[index].IsImmortalThrone = true;
          this.sacks[index].Parse(reader);
        }
        this.itemBlockEnd = (int) reader.BaseStream.Position;
      }
      catch (ArgumentException ex)
      {
        throw;
      }
    }

    public Sack CreateEmptySacks() => new Sack()
    {
      IsModified = false
    };

    public void emptyMainSack()
    {
      this.GetMainSack.EmptySack();
      this.isModified = true;
    }

    public void Save(string fileName, bool backup = false)
    {
      string str1 = "";
      if (backup)
      {
        string str2 = Path.Combine(Path.GetDirectoryName(fileName), "BlaksmithBackup_Player.chr");
        if (!File.Exists(str2))
        {
          Logger.Log("Save file backup created: " + str2);
          File.Copy(fileName, str2);
        }
        str1 = " Original file backed to BlaksmithBackup_Player.chr";
      }
      byte[] buffer = this.Encode();
      using (FileStream fileStream = new FileStream(fileName, FileMode.Create, FileAccess.Write))
        fileStream.Write(buffer, 0, buffer.Length);
      this.isModified = false;
      this.GetMainSack.IsModified = false;
      Logger.Log("Character data saved to " + fileName + str1);
    }

    public byte[] Encode()
    {
      byte[] sourceArray = this.EncodeItemData();
      byte[] destinationArray = new byte[this.itemBlockStart + sourceArray.Length + (this.rawData.Length - this.itemBlockEnd)];
      Array.Copy((Array) this.rawData, 0, (Array) destinationArray, 0, this.itemBlockStart);
      Array.Copy((Array) sourceArray, 0, (Array) destinationArray, this.itemBlockStart, sourceArray.Length);
      Array.Copy((Array) this.rawData, this.itemBlockEnd, (Array) destinationArray, this.itemBlockStart + sourceArray.Length, this.rawData.Length - this.itemBlockEnd);
      return destinationArray;
    }

    private byte[] EncodeEquipmentData()
    {
      int length;
      byte[] buffer;
      using (MemoryStream output = new MemoryStream(2048))
      {
        using (BinaryWriter writer = new BinaryWriter((Stream) output))
        {
          TQData.writeCString(writer, "equipmentCtrlIOStreamVersion");
          writer.Write(this.equipmentCtrlIOStreamVersion);
          this.equipmentSack.Encode(writer);
          length = (int) output.Length;
        }
        buffer = output.GetBuffer();
      }
      if (length == buffer.Length)
        return buffer;
      byte[] destinationArray = new byte[length];
      Array.Copy((Array) buffer, (Array) destinationArray, length);
      return destinationArray;
    }

    private byte[] EncodeItemData()
    {
      int length;
      byte[] buffer;
      using (MemoryStream output = new MemoryStream(2048))
      {
        using (BinaryWriter writer = new BinaryWriter((Stream) output))
        {
          TQData.writeCString(writer, "numberOfSacks");
          writer.Write(this.numberOfSacks);
          TQData.writeCString(writer, "currentlyFocusedSackNumber");
          writer.Write(this.currentlyFocusedSackNumber);
          TQData.writeCString(writer, "currentlySelectedSackNumber");
          writer.Write(this.currentlySelectedSackNumber);
          for (int index = 0; index < this.numberOfSacks; ++index)
            this.sacks[index].Encode(writer);
          length = (int) output.Length;
        }
        buffer = output.GetBuffer();
      }
      if (length == buffer.Length)
        return buffer;
      byte[] destinationArray = new byte[length];
      Array.Copy((Array) buffer, (Array) destinationArray, length);
      return destinationArray;
    }

    public string CharacterName
    {
      get => this.caracterName;
      set
      {
      }
    }

    public string FilePath
    {
      get => this.caracterFile;
      set
      {
      }
    }

    public Sack GetMainSack => this.sacks != null ? this.sacks[0] : (Sack) null;

    public override bool Equals(object obj) => obj != null && obj is Character other && this.Equals(other);

    public override int GetHashCode() => this.caracterName.GetHashCode();

    public bool Equals(Character other) => other != null && this.caracterName.Equals(other.CharacterName);
  }
}
